ouroboros-consensus/ouroboros-consensus-test-infra/src/Test/Util/HardFork/OracularClock.hs

```diff
@@ -0,0 +1,242 @@
+-- exhausted; so the continuation of that call should expect other threads
+-- using the rest of this clock's methods to be stuck (and eg reap them).
+--
+-- Note: though the wallclock-slot correspondence depends on the ledger state,
+-- we have designed our ledgers so that all nodes necessarily use the same
+-- correspondence in the absence of a Common Prefix violation. So this clock
+-- can indeed know /the/ future.
+data OracularClock m = OracularClock
+    { -- | Returns 'False' if the requested slot is already over
+      --
+      -- Note: blocks indefinitely if clock becomes exhausted before requested
+      -- slot.
+      blockUntilSlot :: SlotNo -> m Bool
+
+      -- | The current delay duration until the onset of the next slot
+      --
+      -- Note: blocks indefinitely if the clock is already exhausted.
+    , delayUntilNextSlot :: m NominalDiffTime
+
+      -- | A mock system time
+      --
+      -- Note: this 'BTime.systemTimeCurrent' blocks indefinitely if clock is
+      -- already exhausted.
+    , finiteSystemTime :: BTime.SystemTime m
+
+      -- | Convert 'Control.Monad.Class.MonadTime.getCurrentTime' to 'SlotNo'
+      --
+      -- INVARIANT: In @io-sim@, the @getCurrentTime@ use (by default) the
+      -- monotonic clock, and it advances /before/ threads correspondingly
+      -- wake-up from a @threadDelay@.
+      --
+      -- So we rely on the (monotonic) clock instead of using a ticker thread
+      -- and internal state, since a calling thread and the ticker thread might
+      -- wake up simultaneously from a
+      -- 'Control.Monad.Class.MonadTimer.threadDelay' call: it's ambiguous
+      -- which thread would win the race to respectively read or increment the
+      -- internal state.
+      --
+      -- Note: we do use a ticker thread for 'onSlotChange'.
+      --
+      -- Note: blocks indefinitely if the clock is already exhausted.
+    , getCurrentSlot :: m SlotNo
+
+      -- | See 'onSlotChange'
+    , onSlotChange_ :: HasCallStack
+                    => ResourceRegistry m
+                    -> String
+                    -> (SlotNo -> m ())
+                    -> m (m ())
+
+      -- | Block until the clock is exhausted
+    , waitUntilDone :: m ()
+    }
+
+-- | Forks a thread that executes an action at the onset of each slot
+--
+-- Returns an action that cancels the thread.
+--
+-- INVARIANT: In @io-sim@, there is no race:
+-- @'finiteSystemTime'.systemTimeCurrent@ and hence 'getCurrentSlot' called
+-- from within the given action will always return the correct slot.
+--
+-- See the discussion of ticker threads in 'getCurrentSlot'.
+onSlotChange :: HasCallStack
+             => ResourceRegistry m
+             -> OracularClock m
+             -> String
+             -> (SlotNo -> m ())
+             -> m (m ())
+onSlotChange reg clk = onSlotChange_ clk reg
+    -- jumping the hoop so HasCallStack is useful
+
+-- | Perform a " transaction " within the current slot
```

Comment:
Id maybe avoid the word transaction here, it's (presumably) got nothing to do with UTxO transactions :) (And the signature doesn't ask for an STM transaction either).

Reply:
Also removed (see nearby comment thread).